### ISSUE TITLE
rubgems broken since upgrade to 3.1.2

### ISSUE THREAD

**Issue by Pascal-Vtx** [NixOS/nixpkgs#106545]
**Describe the bug**
Ruby packages (gems) are broken since the commit "ruby: rubygems 3.0.6 -> 3.1.2"  298010982b8986f6669e2b762634dac48253d7ba

It seems that the module autoloading feature of ruby doesn't work anymore for the gems packaged in nixpkgs.

**To Reproduce**
This is a one-line command to test with the gem 'i18n'.
1. with rubygems 3.0.6  ( 788473c is the parent commit of 2980109 )
```sh
$ nix-shell -I 'nixpkgs=https://github.com/NixOS/nixpkgs-channels/archive/788473c6db9c15e2ff6f89287eeee377f2a58a77.tar.gz' -p "ruby.withPackages (ps: [ ps.i18n ] )" --run "ruby -e \"require 'i18n'; print I18n.default_locale;\""
en
```

2. with rubygems 3.1.2 
```sh
$ nix-shell -I 'nixpkgs=https://github.com/NixOS/nixpkgs-channels/archive/298010982b8986f6669e2b762634dac48253d7ba.tar.gz' -p "ruby.withPackages (ps: [ ps.i18n ] )" --run "ruby -e \"require 'i18n'; print I18n.default_locale;\""
Traceback (most recent call last):
        6: from -e:1:in `<main>'
        5: from /nix/store/ygqfnp5qzw0cf7bhski06b2sqbc9wnyi-ruby2.6.5-i18n-0.9.5/lib/ruby/gems/2.6.0/gems/i18n-0.9.5/lib/i18n.rb:39:in `default_locale'
        4: from /nix/store/ygqfnp5qzw0cf7bhski06b2sqbc9wnyi-ruby2.6.5-i18n-0.9.5/lib/ruby/gems/2.6.0/gems/i18n-0.9.5/lib/i18n.rb:26:in `config'
        3: from /nix/store/kd91gyfxgzrprkq8bdaiggnz9a7hr2x7-ruby-2.6.5/lib/ruby/2.6.0/rubygems/core_ext/kernel_require.rb:72:in `require'
        2: from /nix/store/kd91gyfxgzrprkq8bdaiggnz9a7hr2x7-ruby-2.6.5/lib/ruby/2.6.0/rubygems/core_ext/kernel_require.rb:72:in `require'
        1: from /nix/store/d1m9k2vhpgwdrcca4ssckblw484a1453-ruby-gems/lib/ruby/gems/2.6.0/gems/i18n-0.9.5/lib/i18n/config.rb:3:in `<top (required)>'
/nix/store/d1m9k2vhpgwdrcca4ssckblw484a1453-ruby-gems/lib/ruby/gems/2.6.0/gems/i18n-0.9.5/lib/i18n/config.rb:4:in `<module:I18n>': uninitialized constant I18n::Config (NameError)
```

3. with rubygems 3.1.2 and manually loading an internal module of i18n
```sh
$ nix-shell -I 'nixpkgs=https://github.com/NixOS/nixpkgs-channels/archive/298010982b8986f6669e2b762634dac48253d7ba.tar.gz' -p "ruby.withPackages (ps: [ ps.i18n ] )" --run "ruby -e \"require 'i18n/config'; require 'i18n'; print I18n.default_locale;\""
en
```

**Expected behavior**
The commands above should return the same string.
But, it works with rubygems 3.0.6 (step 1.) and fails with 3.1.2 (step 2.) unless the internal module "i18n/config" is manually loaded (step 3.)

**Additional context**
- I took 'i18n' as example but any gem using autoloading should fail in the same way.
- Same issue using bundlerEnv and bundix 
- Same issue with ruby 2.5 or 2.7
- Related issue : #76629

**Metadata**
 - system: `"x86_64-linux"`
 - host os: `Linux 5.4.62, NixOS, 20.03.3318.030e2ce817c (Markhor)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.3.6`
 - channels(root): `"nixos-20.03.3318.030e2ce817c"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`

Maintainer information:
```yaml
attribute: [ ruby, ruby_2_5, ruby_2_6, ruby_2_7 ]
module:
```

**Comment by Pascal-Vtx** [comment]
As temporary workaround I use this overlay to revert rubgems back to 3.0.6 : 
```
    nixpkgs.overlays = [
      (self: super: {
        ruby = let
          rubygems_3_0_6 = (
            import <nixpkgs/pkgs/development/interpreters/ruby/rubygems> {
              stdenv = super.stdenv; lib = super.lib; fetchurl = super.fetchurl; fetchpatch = super.fetchpatch;
            }).overrideAttrs(oldAttrs: rec {
              version = "3.0.6";
              src = super.fetchurl {
                url = "https://rubygems.org/rubygems/rubygems-${version}.tgz";
                sha256 = "1ca1i4xmggizr59m6p28gprlvshczsbx30q8iyzxb2vj4jn8arzx";
              };
              patches = builtins.filter (x: builtins.isPath x) oldAttrs.patches;
            });
        in
        super.ruby.overrideAttrs(oldAttrs: rec {
          postUnpack = ''
          rm -rf $sourceRoot/{lib,test}/rubygems*
          cp -r ${rubygems_3_0_6}/lib/rubygems* $sourceRoot/lib
          cp -r ${rubygems_3_0_6}/test/rubygems $sourceRoot/test
          '';
        });
      })
    ];

```

**Comment by thefloweringash** [comment]
The test case passes for me on current master, so I'm assuming this was solved by #119932. If not, feel free to reopen.
